src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -86,10 +85,9 @@ public static Matcher createMatcher(RubyContext context, RubyRegexp regexp, Rope
             regex = encodingCache.getOrCreate(enc, e -> makeRegexpForEncoding(context, regexp, e, currentNode));
         }
 
-        return regex.matcher(stringBytes, start, stringBytes.length);
+        return getMatcher(regex, stringBytes, start);
     }
```

Comment:
The change will cause two boundaries (getOrCreate() and getMatcher()) instead of 1, but the first one might not be executed if `regex.getEncoding() == enc`.
Avoiding the boundary here might allow to escape analyze the `Rope stringRope`, if it already has its `byte[]` computed.

It seems fine, but I think we should inline `createMatcher` into its only caller, and based on that profile the `regex.getEncoding() != enc` condition.
I think we should also turn `checkEncoding` into a node.

Reply:
I was just looking at turning `checkEncoding` into a node. If not a problem, I'll handle that in a separate PR.